File: app/src/main/java/com/example/server/ServerSorter.kt
package com.example.server

import android.util.Log
import kotlin.math.abs

class ServerSorter(
    private val weights: SortingWeights = SortingWeights()
) {

    fun sort(
        streams: List<ServerStream>,
        customWeights: SortingWeights? = null
    ): List<ServerStream> {
        val w = customWeights ?: weights

        val sorted = streams.sortedWith { a, b ->
            // 1. Playable availability first
            if (a.isPlayable != b.isPlayable) {
                return@sortedWith if (b.isPlayable) 1 else -1
            }

            // Calculate composite ranking score
            // Quality component: 0 to 400 scaled by qualityWeight
            val aQualityScore = (a.quality.rank.toDouble() / 400.0) * w.qualityWeight
            val bQualityScore = (b.quality.rank.toDouble() / 400.0) * w.qualityWeight

            // Speed component: lower responseTime is better, scaled by speedWeight
            // 100ms -> score 1.0, 3000ms -> score 0.0
            val aSpeedScore = (1.0 - (a.responseTimeMs.toDouble() / 3000.0).coerceIn(0.0, 1.0)) * w.speedWeight
            val bSpeedScore = (1.0 - (b.responseTimeMs.toDouble() / 3000.0).coerceIn(0.0, 1.0)) * w.speedWeight

            // Priority component: higher priority is better, scaled by priorityWeight
            val aPriorityScore = (a.providerPriority.coerceIn(1, 100).toDouble() / 100.0) * w.priorityWeight
            val bPriorityScore = (b.providerPriority.coerceIn(1, 100).toDouble() / 100.0) * w.priorityWeight

            val aTotalScore = aQualityScore + aSpeedScore + aPriorityScore
            val bTotalScore = bQualityScore + bSpeedScore + bPriorityScore

            // Compare by total score
            val scoreDiff = bTotalScore - aTotalScore
            if (abs(scoreDiff) > 1.0) {
                return@sortedWith if (scoreDiff > 0) 1 else -1
            }

            // Tie-breaking: Quality -> Speed -> Server Priority
            if (a.quality.rank != b.quality.rank) {
                return@sortedWith b.quality.rank.compareTo(a.quality.rank)
            }
            if (abs(a.responseTimeMs - b.responseTimeMs) > 100) {
                return@sortedWith a.responseTimeMs.compareTo(b.responseTimeMs)
            }
            a.serverPriority.compareTo(b.serverPriority)
        }

        // Tag the top valid stream as recommended
        var recommendedTagged = false
        val finalStreams = sorted.map { stream ->
            if (!recommendedTagged && stream.isPlayable) {
                recommendedTagged = true
                stream.copy(isRecommended = true)
            } else {
                stream.copy(isRecommended = false)
            }
        }

        val playableCount = finalStreams.count { it.isPlayable }
        Log.d("ServerSorter", "[Sorter] $playableCount playable servers sorted (total: ${finalStreams.size})")

        return finalStreams
    }
}

File: src/services/server/ServerSorter.ts
import { ServerStream, StreamQuality } from './ServerTypes';

const QUALITY_SCORES: Record<StreamQuality, number> = {
    '4K UHD': 400,
    '1080p FHD': 300,
    '720p HD': 200,
    '480p SD': 100,
    'Auto': 250
};

export class ServerSorter {
    /**
     * Sorts servers according to:
     * 1. Playable state (playable first)
     * 2. Quality rank (higher resolution first)
     * 3. Response latency (faster response time first)
     * 4. Provider priority
     * 5. Server priority
     */
    sort(streams: ServerStream[]): ServerStream[] {
        const sorted = [...streams].sort((a, b) => {
            // 1. Playability
            if (a.isPlayable !== b.isPlayable) {
                return (b.isPlayable ? 1 : 0) - (a.isPlayable ? 1 : 0);
            }

            // 2. Quality
            const qualityA = QUALITY_SCORES[a.quality] || 200;
            const qualityB = QUALITY_SCORES[b.quality] || 200;
            if (qualityA !== qualityB) {
                return qualityB - qualityA;
            }

            // 3. Response latency (lower ms is better)
            const latencyA = a.responseTimeMs ?? 9999;
            const latencyB = b.responseTimeMs ?? 9999;
            if (Math.abs(latencyA - latencyB) > 200) { // meaningful latency difference
                return latencyA - latencyB;
            }

            // 4. Provider priority (lower number = higher priority)
            if (a.providerPriority !== b.providerPriority) {
                return a.providerPriority - b.providerPriority;
            }

            // 5. Server internal priority
            return a.serverPriority - b.serverPriority;
        });

        // Tag the first playable stream as recommended
        let assigned = false;
        return sorted.map((s, index) => {
            if (!assigned && (s.isPlayable ?? true)) {
                assigned = true;
                return { ...s, isRecommended: true };
            }
            return { ...s, isRecommended: false };
        });
    }
}
